File: WebGL/particle_terrain.js
function ParticleTerrain( w ) {
	this.width = w;
	this.data  = new Array( this.width * this.width );
	
	for( var i = 0; i < this.data.length; i++ ) {
		this.data[i] = 0.0;
	}	
}

ParticleTerrain.prototype = {
	dir: function() {
		return this.randInt( 3 );
	},
	
	generate: function( land_radius, num_islands, iterations, sm ) {
		
		// Displacement. Increase if we want taller land masses
		var disp = 3;
		
		for( var j = 0; j < num_islands; j++ ) {

			// Find a random spot to grow an island
			var sx = this.randInt( this.width - 1 );
			var sy = this.randInt( this.width - 1 );

			var x = sx, y = sy;
			for( var i = 0; i < iterations; i++ ) {
				
				var d = this.getData( x, y );
				// Check neighbors
				if( this.getData( x-1, y ) < d ) {
					this.setData( x-1, y, this.getData( x-1, y ) + disp );
				} else if( this.getData( x+1, y ) < d ) {
					this.setData( x+1, y, this.getData( x+1, y ) + disp );
				} else if( this.getData( x, y-1 ) < d ) {
					this.setData( x, y-1, this.getData( x, y-1 ) + disp );
				} else if( this.getData( x, y+1 ) < d ) {
						this.setData( x, y+1, this.getData( x, y+1 ) + disp );
				} else {
					this.setData( x, y, d + disp );
				}

				switch( this.dir() ) {
					case 0:
						y++;
						if( this.inCircle( sx, sy, x, y, land_radius ) && 
								y > 0 && y < this.width ) {
							break;
						} else {
							y--;
						}
					case 1:
						y--;
						if( this.inCircle( sx, sy, x, y, land_radius ) && 
								y > 0 && y < this.width ) {
							break;
						} else {
							y++;
						}
					case 2:
						x++;
						if( this.inCircle( sx, sy, x, y, land_radius ) && 
								x > 0 && x < this.width ) {
							break;
						} else {
							x--;
						}
					case 3:
						x--;
						if( this.inCircle( sx, sy, x, y, land_radius ) &&
								x > 0 && x < this.width ) {
							break;
						} else {
							x++;
						}
				}
			}
		}

		this.smooth( sm );
		return this.data;
	},
	
	getData: function( x, y ) {
		return this.data[ x * this.width + y ];
	},
	
	inCircle: function( sx, sy, x, y, r ) {
		return ( Math.pow( x-sx, 2) + Math.pow( y-sy, 2 ) ) < Math.pow( r, 2 );	
	},
	
	// Returns a number between 0 and num, inclusive.
	randInt: function( num ) {
		return Math.floor( Math.random()*(num+1) );	
	},
	
	setData: function( x, y, val ) {
		this.data[ x * this.width + y ] = val;
	},

	smooth: function( k ) {
		for( var x = 1; x < this.width; x++ ) {
			for( var z = 0; z < this.width; z++ ) {
				this.data[ x * this.width + z ] = 
				this.data[ (x-1) * this.width + z ] * ( 1 - k ) + 
				this.data[ x * this.width + z ] * k;
			}
		}

		for( var x = this.width-2; x >= 0; x-- ) {
			for( var z = 0; z < this.width; z++ ) {
				this.data[ x * this.width + z ] = 
				this.data[ (x+1) * this.width + z ] * ( 1 - k ) + 
				this.data[ x * this.width + z ] * k;
			}
		}

		for( var x = 0; x < this.width; x++ ) {
			for( var z = this.width-2; z >= 0; z-- ) {
				this.data[ x * this.width + z ] = 
				this.data[ x * this.width + (z+1) ] * ( 1 - k ) + 
				this.data[ x * this.width + z ] * k;
			}
		}

		for( var x = 0; x < this.width; x++ ) {
			for( var z = 1; z < this.width; z++ ) {
				this.data[ x * this.width + z ] = 
				this.data[ x * this.width + (z-1) ] * ( 1 - k ) + 
				this.data[ x * this.width + z ] * k;
			}
		}		
	}
};

File: CPlusPlus/ParticleTerrain.cpp
/*
 *  ParticleTerrain.cpp
 *  FinalProject
 *
 *  Created by Andrew Huynh on 12/6/09.
 *  Copyright 2009 ATHLabs. All rights reserved.
 *
 */

#include "ParticleTerrain.h"
#include <stdlib.h>
#include <math.h>
#include <stdio.h>

ParticleTerrain::ParticleTerrain( int w ) {
	width = w;

	data = (float*)malloc( width * width * sizeof(float) );
	
	// It's all water for now.
	for( int i = 0; i < width * width; i++ ) {
		data[i] = -7.0f;
	}
}

float ParticleTerrain::getData( int x, int y ) {
	return data[x * width + y];
}

void ParticleTerrain::setData( int x, int y, float val ) {
	data[ x * width + y ] = val;
}

int ParticleTerrain::getWidth() {
	return width;
}

bool ParticleTerrain::inCircle( int sx, int sy, int x, int y, int r ) {
	return ( pow( x-sx, 2 ) + pow( y-sy, 2 ) ) < pow( r, 2 );
}

void ParticleTerrain::generate( int seed, int land_radius, int num_islands, int iterations, float sm ) {
	k = sm;
	srand( seed );
	
	for( int j = 0; j < num_islands; j++ ) {
		int sx = rand() % (width-1) + 1;
		int sy = rand() % (width-1) + 1;
	
		int x = sx;
		int y = sy;
		
		for( int i = 0; i < iterations; i++ ) {
		
			int d = getData( x, y );
			int disp = 3;//genDisp( 10 );

			// Check neighbors
			if( getData( x-1, y ) < d ) {
				setData( x-1, y, getData( x-1, y ) + disp );
			} else if( getData( x+1, y ) < d ) {
				setData( x+1, y, getData( x+1, y ) + disp );
			} else if( getData( x, y-1 ) < d ) {
				setData( x, y-1, getData( x, y-1 ) + disp );
			} else if( getData( x, y+1 ) < d ) {
				setData( x, y+1, getData( x, y+1 ) + disp );
			} else {
				setData( x, y, d + disp );
			}
			
			int step = dir();
			switch( step ) {
				case 0:
					y++;
					if( inCircle( sx, sy, x, y, land_radius ) && 
							y > 0 && y < width ) {
						break;
					} else {
						y--;
					}
				case 1:
					y--;
					if( inCircle( sx, sy, x, y, land_radius ) && 
							y > 0 && y < width ) {
						break;
					} else {
						y++;
					}
				case 2:
					x++;
					if( inCircle( sx, sy, x, y, land_radius ) && 
							x > 0 && x < width ) {
						break;
					} else {
						x--;
					}
				case 3:
					x--;
					if( inCircle( sx, sy, x, y, land_radius ) &&
							x > 0 && x < width ) {
						break;
					} else {
						x++;
					}					
			}
		}
	}
	
	smooth();
}

int ParticleTerrain::dir() {
	return rand() % 4;
}

float ParticleTerrain::genDisp( float maxDisp ) {
	return (2*maxDisp) * ((float)rand()/RAND_MAX) - maxDisp;
}

void ParticleTerrain::smooth() {
	for( int x = 1; x < width; x++ ) {
		for( int z = 0; z < width; z++ ) {
			data[ x * width + z ] = 
				data[ (x-1) * width + z ] * ( 1 - k ) + 
				data[ x * width + z ] * k;
		}
	}

	for( int x = width-2; x >= 0; x-- ) {
		for( int z = 0; z < width; z++ ) {
			data[ x * width + z ] = 
				data[ (x+1) * width + z ] * ( 1 - k ) + 
				data[ x * width + z ] * k;
		}
	}
	
	for( int x = 0; x < width; x++ ) {
		for( int z = width-2; z >= 0; z-- ) {
			data[ x * width + z ] = 
				data[ x * width + (z+1) ] * ( 1 - k ) + 
				data[ x * width + z ] * k;
		}
	}

	for( int x = 0; x < width; x++ ) {
		for( int z = 1; z < width; z++ ) {
			data[ x * width + z ] = 
				data[ x * width + (z-1) ] * ( 1 - k ) + 
				data[ x * width + z ] * k;
		}
	}
}